client/starwhale/api/_impl/dataset/builder.py

```diff
@@ -210,271 +242,202 @@ class SWDSBinBuildExecutor(BaseBuildExecutor):
+            bin2link_threshold: int = D_BIN_TO_LINK_THRESHOLD,
+        ) -> None:
+            self.tabular_dataset = tabular_dataset
+            self.work_dir = work_dir
+            self.data_output_dir = data_output_dir
+            self.volume_bytes_size = volume_bytes_size
+            self.bin2link_threshold = bin2link_threshold
+            self.alignment_bytes_size = alignment_bytes_size
+            self.ds_copy_candidates: t.Dict[str, SWDSBinBuildExecutor._SrcPathSpec] = {}
+            self.wrote_size = 0
+            _, bin_writer_path = tempfile.mkstemp(
+                prefix="bin-writer-", dir=str(self.work_dir.absolute())
             )
-
-            _artifact: BaseArtifact
-            if isinstance(row_data, bytes):
-                _artifact = Binary(row_data, self.default_data_mime_type)
-            elif isinstance(row_data, BaseArtifact):
-                _artifact = row_data
-            else:
-                raise NoSupportError(f"data type {type(row_data)}")
-
-            if not dataset_annotations:
-                # TODO: check annotations type and name
-                dataset_annotations = row_annotations
-
-            _bin_section = self._write(dwriter, _artifact.to_bytes())
-            self.tabular_dataset.put(
-                TabularDatasetRow(
-                    id=idx,
-                    data_link=Link(str(fno)),
-                    data_format=self.data_format_type,
-                    object_store_type=ObjectStoreType.LOCAL,
-                    data_offset=_bin_section.raw_data_offset,
-                    data_size=_bin_section.raw_data_size,
-                    _swds_bin_offset=_bin_section.offset,
-                    _swds_bin_size=_bin_section.size,
-                    data_origin=DataOriginType.NEW,
-                    data_type=_artifact.astype(),
-                    annotations=row_annotations,
-                    _append_seq_id=append_seq_id,
-                )
+            self.dwriter_path = Path(bin_writer_path)
+            self.dwriter = self.dwriter_path.open("wb")
+            self.total_bin_size = 0
+            self._to_update_rows: t.List[TabularDatasetRow] = []
+            self._lock = threading.Lock()
+
+        def __enter__(self) -> SWDSBinBuildExecutor._BinWriter:
+            return self
+
+        def __exit__(
+            self,
+            type: t.Optional[t.Type[BaseException]],
+            value: t.Optional[BaseException],
+            trace: TracebackType,
+        ) -> None:
+            if value:  # pragma: no cover
+                logger.warning(f"type:{type}, exception:{value}, traceback:{trace}")
+
+            self.close()
+
+        def write_row(self, row: TabularDatasetRow) -> None:
+            with self._lock:
+                artifacts = row.artifacts()
+                artifacts_with_bin = False
+                for v in artifacts:
+                    if not v.link and isinstance(v.fp, (str, Path)):
```

Comment:
there three if condions are mutual exclusion, use if-elif-else structure?

Reply:
Merge 301-303 into one condition for local file case?

Reply:
Can str type infer file path localtion？

Reply:
> Can str type infer file path localtion？

If the name  `fp` is shot for file path, I think we could consider str as the location

Reply:
> Merge 301-303 into one condition for local file case?

line 303 and 301 are parallel conditions。 301 convert str v.fp to v.link if v.fp is str. 303 judges the v.link.

Reply:
> there three if condions are mutual exclusion, use if-elif-else structure?

ok